### DISCUSSION TITLE
Duplicated `/opt/homebrew/bin` entry in PATH

### DISCUSSION THREAD

**Discussion by Tobbe** [Homebrew/discussions#5502] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.3.9
ORIGIN: https://github.com/Homebrew/brew
HEAD: 1c98a33a376e5b35ebc801d0f7d6ba7b9eb118ac
Last commit: 10 days ago
Core tap JSON: 18 Jul 10:25 UTC
Core cask tap JSON: 18 Jul 10:25 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 3.3.3 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.3/bin/ruby
CPU: dodeca-core 64-bit arm_lobos
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 14.1.1-arm64
CLT: 15.0.0.0.1.1694021235
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Settings.

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 15.4.


Warning: You have the following deprecated, official taps tapped:
  Homebrew/homebrew-cask-fonts
Untap them with `brew untap`.

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected header files:
  /usr/local/include/node/cppgc/allocation.h
  /usr/local/include/node/cppgc/common.h
# Very long list...
  /usr/local/include/node/zconf.h
  /usr/local/include/node/zlib.h
```

### Description of issue

When I run `echo $PATH | sed 's/:/\n/g'` I get this output:

```
/opt/homebrew/bin
/opt/homebrew/sbin
/usr/local/bin
/System/Cryptexes/App/usr/bin
/usr/bin
/bin
/usr/sbin
/sbin
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin
/usr/local/go/bin
/opt/homebrew/bin
/opt/homebrew/sbin
```

As you can see I have `/opt/homebrew/bin` (and sbin) in there twice. I think it should only be added once.

I can edit `/etc/zprofile` to disable `path_helper`; like this:
```
# System-wide profile for interactive zsh(1) login shells.

# Setup user specific overrides for this in ~/.zprofile. See zshbuiltins(1)
# and zshoptions(1) for more details.

# if [ -x /usr/libexec/path_helper ]; then
#         eval `/usr/libexec/path_helper -s`
# fi
```
If I do that I only get the homebrew entries once in my PATH
```
/opt/homebrew/bin
/opt/homebrew/sbin
/usr/bin
/bin
/usr/sbin
/sbin
```

But obviously I don't want to have to do that. I need the other paths I get from running `path_helper`

Here's some more debugging info (`/etc/paths` and `/etc/paths.d/`):
```
tobbe@Tobbes-MacBook-Pro14 ~ % cat /etc/paths
/usr/local/bin
/System/Cryptexes/App/usr/bin
/usr/bin
/bin
/usr/sbin
/sbin
tobbe@Tobbes-MacBook-Pro14 ~ % ls -la /etc/paths.d
total 16
drwxr-xr-x   4 root  wheel   128 Feb  9 15:19 .
drwxr-xr-x  81 root  wheel  2592 Jul 18 13:34 ..
-rw-r--r--   1 root  wheel   224 Nov  3  2023 10-cryptex
-rw-r--r--   1 root  wheel    18 Feb  2 20:16 go
```

I found this issue from last year https://github.com/Homebrew/brew/issues/14560 that suggests running `script /tmp/zsh.log zsh -xl`. Doing that, and then looking at the logs I get (slightly trimmed in the middle): 
```
tobbe@Tobbes-MacBook-Pro14 rw-rsc-status % cat /tmp/zsh.log           
Script started on Thu Jul 18 13:43:56 2024
Command: zsh -xl
+/etc/zprofile:6> [ -x /usr/libexec/path_helper ']'
+/etc/zprofile:7> /usr/libexec/path_helper -s
+/etc/zprofile:7> eval 'PATH="/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/usr/local/go/bin:/opt/homebrew/bin:/opt/homebrew/sbin";' export 'PATH;' 'MANPATH="/usr/share/man:/usr/local/share/man:/opt/homebrew/share/man:";' export 'MANPATH;'
+(eval):1> PATH=/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/usr/local/go/bin:/opt/homebrew/bin:/opt/homebrew/sbin 
+(eval):1> export PATH
+(eval):1> MANPATH=/usr/share/man:/usr/local/share/man:/opt/homebrew/share/man: 
+(eval):1> export MANPATH
+/Users/tobbe/.zprofile:2> /opt/homebrew/bin/brew shellenv
+/Users/tobbe/.zprofile:2> eval $'export HOMEBREW_PREFIX="/opt/homebrew";\nexport HOMEBREW_CELLAR="/opt/homebrew/Cellar";\nexport HOMEBREW_REPOSITORY="/opt/homebrew";\nexport PATH="/opt/homebrew/bin:/opt/homebrew/sbin${PATH+:$PATH}";\nexport MANPATH="/opt/homebrew/share/man${MANPATH+:$MANPATH}:";\nexport INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";'
+(eval):1> export HOMEBREW_PREFIX=/opt/homebrew
+(eval):2> export HOMEBREW_CELLAR=/opt/homebrew/Cellar
+(eval):3> export HOMEBREW_REPOSITORY=/opt/homebrew
+(eval):4> export PATH=/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/usr/local/go/bin:/opt/homebrew/bin:/opt/homebrew/sbin
+(eval):5> export MANPATH=/opt/homebrew/share/man:/usr/share/man:/usr/local/share/man:/opt/homebrew/share/man::
+(eval):6> export INFOPATH=/opt/homebrew/share/info:/opt/homebrew/share/info:/opt/homebrew/share/info:
+/etc/zshrc:7> [[+/etc/zshrc:7> locale LC_CTYPE
+/etc/zshrc:7> [[ UTF-8 == UTF-8 ]]
+/etc/zshrc:8> setopt COMBINING_CHARS
+/etc/zshrc:12> disable log
+/etc/zshrc:15> HISTFILE=/Users/tobbe/.zsh_history 
+/etc/zshrc:16> HISTSIZE=2000 
+/etc/zshrc:17> SAVEHIST=1000 
+/etc/zshrc:20> setopt BEEP
+/etc/zshrc:24> [[ -r /Users/tobbe/.zkbd/xterm-256color-apple ]]
+/etc/zshrc:27> typeset -g -A key
+/etc/zshrc:29> [[ -n $'\C-[OP' ]]
+/etc/zshrc:29> key[F1]=$'\C-[OP' 

# trimmed more key stuff here

+/etc/zshrc:67> [[ -n $'\C-[OB' ]]
+/etc/zshrc:67> bindkey $'\C-[OB' down-line-or-search
+/etc/zshrc:70> PS1='%n@%m %1~ %# ' 
+/etc/zshrc:73> [ -r /etc/zshrc_vscode ']'
tobbe@Tobbes-MacBook-Pro14 rw-rsc-status % 

Command exit status: 1
Script done on Thu Jul 18 13:43:57 2024
```

If I change my `~/.zprofile` to look like this I don't get any homebrew paths at all in PATH
```
tobbe@Tobbes-MacBook-Pro14 ~ % cat .zprofile

# eval "$(/opt/homebrew/bin/brew shellenv)"
```

What can I do to only have the homebrew paths show up once in my PATH?

**Comment by SMillerDev** [comment]
What does `cat /etc/paths.d/go` say? But also, what does the duplicate entry impact specifically? You shell should be able to handle that without any issues.

**Reply by Tobbe** [reply]
@SMillerDev Thanks for your reply

```
tobbe@Tobbes-MacBook-Pro14 ~ % cat /etc/paths.d/go
/usr/local/go/bin
```

> But also, what does the duplicate entry impact specifically? You shell should be able to handle that without any issues.

It's not my shell that's the issue. It's some other tool that gets confused when `which -a` returns more than one result with identical paths

And not that I think it matters, but this is the content of go/bin

```
tobbe@Tobbes-MacBook-Pro14 ~ % ls -la /usr/local/go/bin
total 29288
drwxr-xr-x   4 root  wheel       128 Feb  2 20:19 .
drwxr-xr-x  18 root  wheel       576 Feb  2 20:16 ..
-rwxr-xr-x   1 root  wheel  12370000 Feb  2 20:19 go
-rwxr-xr-x   1 root  wheel   2617408 Feb  2 20:19 gofmt
```

(Also wanted to make sure you saw my latest edit to my question about commenting out the eval in .zprofile)

**Comment by Tobbe** [comment]
Running `/opt/homebrew/bin/brew shellenv` prints the following to my console
```
export HOMEBREW_PREFIX="/opt/homebrew";
export HOMEBREW_CELLAR="/opt/homebrew/Cellar";
export HOMEBREW_REPOSITORY="/opt/homebrew";
export PATH="/opt/homebrew/bin:/opt/homebrew/sbin${PATH+:$PATH}";
export MANPATH="/opt/homebrew/share/man${MANPATH+:$MANPATH}:";
export INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";
```

Having that in my `.zprofile` instead still gives me the double paths
```
# eval "$(/opt/homebrew/bin/brew shellenv)"

export HOMEBREW_PREFIX="/opt/homebrew";
export HOMEBREW_CELLAR="/opt/homebrew/Cellar";
export HOMEBREW_REPOSITORY="/opt/homebrew";
export PATH="/opt/homebrew/bin:/opt/homebrew/sbin${PATH+:$PATH}";
export MANPATH="/opt/homebrew/share/man${MANPATH+:$MANPATH}:";
export INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";
```

Commenting out all of them I get no homebrew paths at all in PATH. Same as when all I had in there was the commented out `eval "..."` command.

So far nothing too surprising 🙂 

Now things get more interesting. 

This whole thing only seems to be a problem in my editor's built-in terminals❗❗ 
I get the same behavior in both VSCode and Zed.
But in iTerm2 I actually have no problem with duplicated paths.

With that in mind I only commented out the `export PATH=...` line and kept all the other `export ...` lines enabled.
Now, in iTerm2, I still don't get any homebrew paths in PATH
But in VSCode I now get them once.

And on close inspection there are more PATH entries that get duplicated in the VSCode and Zed terminals, that don't get duplicated in iTerm2 or Terminal

So I'm going to let homebrew off the the hook here 😀 and just assume this is an issue with VSCode and Zed

**Reply by gromgit** [reply]
> Running `/opt/homebrew/bin/brew shellenv` prints the following to my console
> 
> ```
> export HOMEBREW_PREFIX="/opt/homebrew";
> export HOMEBREW_CELLAR="/opt/homebrew/Cellar";
> export HOMEBREW_REPOSITORY="/opt/homebrew";
> export PATH="/opt/homebrew/bin:/opt/homebrew/sbin${PATH+:$PATH}";
> export MANPATH="/opt/homebrew/share/man${MANPATH+:$MANPATH}:";
> export INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";
> ```

This is expected behaviour. `brew shellenv` will churn out the necessary `export`s if it doesn't see Homebrew's `bin` and `sbin` at the start of your `PATH`.

> Having that in my `.zprofile` instead still gives me the double paths
> 
> ```
> # eval "$(/opt/homebrew/bin/brew shellenv)"
> 
> export HOMEBREW_PREFIX="/opt/homebrew";
> export HOMEBREW_CELLAR="/opt/homebrew/Cellar";
> export HOMEBREW_REPOSITORY="/opt/homebrew";
> export PATH="/opt/homebrew/bin:/opt/homebrew/sbin${PATH+:$PATH}";
> export MANPATH="/opt/homebrew/share/man${MANPATH+:$MANPATH}:";
> export INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";
> ```

If that's literally what's in your `.zprofile`, that's not what the installer asked you to do. Replace all that with:
```
eval "$(/opt/homebrew/bin/brew shellenv)"
```

> This whole thing only seems to be a problem in my editor's built-in terminals❗❗ I get the same behavior in both VSCode and Zed. But in iTerm2 I actually have no problem with duplicated paths.

Then it's likely you're resetting PATH in your VSCode and Zed configs.

**Reply by Tobbe** [reply]
> If that's literally what's in your .zprofile, that's not what the installer asked you to do

I just tried that as an experiment. What I had before, and what I have now again after my experiment, is only 
```

eval "$(/opt/homebrew/bin/brew shellenv)"
```

**Comment by meichuan** [comment]
Hope it's not too late. :-) Happened to encounter this problem today. Thought maybe I had a clue to what really happened. May I suggest a rather simple trick to "solve" your problem? Just remove the `homebrew` file under `/etc/paths.d` after configuring `$HOME/.zprofile` according to the installation instruction. Voila!
